src/services/UnderlyingYieldService.ts

```diff
@@ -0,0 +1,376 @@
+        return {
+          rate: event.args['totalEth'] / event.args['rethSupply'],
+          timestamp: event.args['blockTimestamp'],
+        };
+      })
+      .filter((rate) => rate !== null) as LstRate[];
+    if (rates === null || rates.length < 2) {
+      return await getApyFromApi();
+    } else {
+      const apy = getApyFromLstRates(rates[rates.length - 1], rates[0], 365); // rewards are distributed approximately every 24 hours: (source: https://docs.rocketpool.net/guides/staking/overview#the-reth-token)
+      return apy;
+    }
+  };
+
+  getEthxAPY = async (provider: Provider, currentBlockNumber: number) => {
+    const getApyFromApi = async () => {
+      const res = await fetch('https://universe.staderlabs.com/eth/apy');
+      const resParsed: {
+        value: number;
+      } = await res.json();
+      return resParsed.value / 100;
+    };
+
+    const abi = [
+      {
+        anonymous: false,
+        inputs: [
+          { indexed: false, internalType: 'uint256', name: 'block', type: 'uint256' },
+          { indexed: false, internalType: 'uint256', name: 'totalEth', type: 'uint256' },
+          { indexed: false, internalType: 'uint256', name: 'ethxSupply', type: 'uint256' },
+          { indexed: false, internalType: 'uint256', name: 'time', type: 'uint256' },
+        ],
+        name: 'ExchangeRateUpdated',
+        type: 'event',
+      },
+    ];
+
+    const contract = new Contract('0xF64bAe65f6f2a5277571143A24FaaFDFC0C2a737', abi); // Stader Labs Oracle
+    const connectedContract = contract.connect(provider);
+
+    const events = await connectedContract.queryFilter(
+      connectedContract.filters.ExchangeRateUpdated(),
+      currentBlockNumber - BLOCKS_A_DAY * EVENTS_PERIOD_DAYS,
+      currentBlockNumber
+    );
+
+    const rates = events
+      .map((event) => {
+        if (!event.args || !event.args['totalEth'] || !event.args['ethxSupply']) return null;
+        return {
+          rate: event.args['totalEth'] / event.args['ethxSupply'],
+          timestamp: event.args['time'],
+        };
+      })
+      .filter((rate) => rate !== null) as LstRate[];
+    if (rates === null || rates.length < 2) {
+      return await getApyFromApi();
+    } else {
+      const apy = getApyFromLstRates(rates[rates.length - 1], rates[0], 365); // rewards seems to be distributed every 24 hours
+      return apy;
+    }
+  };
+
+  getCbethAPY = async (provider: Provider, currentBlockNumber: number) => {
+    const abi = [
+      {
+        anonymous: false,
+        inputs: [
+          { indexed: true, internalType: 'address', name: 'oracle', type: 'address' },
+          { indexed: false, internalType: 'uint256', name: 'newExchangeRate', type: 'uint256' },
+        ],
+        name: 'ExchangeRateUpdated',
+        type: 'event',
+      },
+    ];
+
+    const contract = new Contract('0x9b37180d847B27ADC13C2277299045C1237Ae281', abi); // cbETH Oracle
```

Comment:
all contract addresses should be pulled from the aave-address-book. If they are not they should be added https://github.com/bgd-labs/aave-address-book

Reply:
Even for "tricky" contracts like RocketNetworkBalances (`0x6Cc65bF618F55ce2433f9D8d827Fc44117D81399`)? (contract that has no relation with Aave?

Reply:
I've just push a [PR](https://github.com/bgd-labs/aave-address-book/pull/481)
I'll tag you once the modifications will be done on this PR

Reply:
Chiming in here as a maintainer of address-book.

We don't really want to list these addresses on address-book. The reason is that address book first and foremost is a library for usage in solidity:
> This repository contains an up-to-date registry of all addresses of the Aave ecosystem's smart contracts, for its usage in Solidity codebases.

and the js export is a low effort gimmick we added at a later point to also expose these addresses in js. This by no means implies though, that we'll add any address that might be useful in js on the address-book as this always comes at a cost for the solidity users of the library.

---

This out of the way, i had a quick look at what is done here and want to share my 2ct:

1. I think what @MartinGbz did here is quite valuable and could be useful for other ecosystem projects. I'd like to encourage you to develop this as an independent & modular package so you can rely on viem, multicall etc. We'd be happy to help & contribute.
2. probably it would be better to use `getLogs` with a topics filter over `queryFilter`. The reason is that queryFilter is not as widely supported and buggy on some node implementations.
3. For use in an ui (like this one) I think it's quite wasteful to actually do what is done here. Every user will need to do multiple api calls, fetch events over time and do calculations etc. Offloading unnecessary work to the client comes as a big cost. I would highly encourage the maintainers of this app to put 1) behind some cached api so users have to do a single request and not calculate anything clientside.